src/Ethereum/ABI/ParamNumber.cpp

```diff
@@ -0,0 +1,63 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "ParamNumber.h"
+#include "Util.h"
+
+#include "../../Data.h"
+#include "../../uint256.h"
+
+#include <string>
+#include <cassert>
+
+using namespace TW;
+using namespace TW::Ethereum::ABI;
+
+void ParamUIntN::setVal(uint256_t val) {
+    // mask it to the given bits
+    _val = val & _mask;
+}
+
+bool ParamUIntN::decode(const Data& encoded, size_t& offset_inout) {
+    uint256_t temp;
+    auto res = decodeNumber(encoded, temp, offset_inout);
+    setVal(temp);
+    return res;
+}
+
+void ParamUIntN::init() {
+    assert(bits >= 8 && bits <= 256 && (bits % 8) == 0);
+    _mask = (uint256_t(1) << bits) - 1;
+}
+
+void ParamIntN::setVal(int256_t val) {
+    // mask it to the given bits
+    if (val < 0) {
+        _val = Util::int256FromUint256(~((~((uint256_t)val)) & _mask));
+    } else {
+        _val = Util::int256FromUint256(((uint256_t)val) & _mask);
+    }
+}
+
+bool ParamIntN::decodeNumber(const Data& encoded, int256_t& decoded, size_t& offset_inout) {
+    uint256_t valU;
+    auto res = TW::Ethereum::ABI::decode(encoded, valU, offset_inout);
+    decoded = Util::int256FromUint256(valU);
+    return res;
+}
+
+bool ParamIntN::decode(const Data& encoded, size_t& offset_inout) {
+    int256_t temp;
+    auto res = decodeNumber(encoded, temp, offset_inout);
+    setVal(temp);
+    return res;
+}
+
+void ParamIntN::init()
+{
+    assert(bits >= 8 && bits <= 256 && (bits % 8) == 0);
```

Comment:
@hewigovens is it safe to use assert in production?

Reply:
assert won't work in release, the app needs to prepare the data correctly

Reply:
Assert have to be used with care.  They are optimized out from release build.  I use it only in fatal, programming error, 'never can happen' cases.  Sometimes I use them in incorrect input case, but also handling it nicely, like this:
    assert(param != nullptr);
    if (param == nullptr) { return -1; }
This helps catching problems in debug/tests.